torch/csrc/autograd/python_engine.cpp

```diff
@@ -237,7 +238,13 @@ PyObject *THPEngine_run_backward(PyObject *self, PyObject *args, PyObject *kwarg
       THPUtils_assert(input_var->cdata.requires_grad(),
           "One of the differentiated Tensors does not require grad");
       if (!grad_fn) {
-        output_edges.emplace_back();
+        // Since input has no grad_accumulator, its guaranteed to be not reachable.
+        // We create a dummy edge with no next edges and a non-nullptr grad_fn so
+        // nodes (e.g., mul when an operand is scalar) that have invalid next_edges
+        // don't get erroneously assigned `needed = True` in exec_info. Note that this
+        // doesn't happen in the cpp api because scalars are only wrapped into tensors
```

Comment:
Not sure this is true when you mix tensors that do and don't require gradients, see above.

Reply:
Ahh, forgot about this case. Good catch!